GoogleForCreators/web-stories-wp
# Taxonomy: Check capabilities for managing and assigning terms

## Context

Not all users have the capability to add new categories in taxonomies. This limits that. 

## Summary

(Revised after Pascal's [comment](https://github.com/google/web-stories-wp/pull/9099#issuecomment-924398219)/[contribution](https://github.com/google/web-stories-wp/pull/9099/commits/974ac11a7b76c34c66d4c4cc6a9e8241d49029d2))

Taxonomy capabilities for managing and assigning terms are listed in the story response under `capabilities`, they are given based on the taxonomy `slug` and prefixed by either `create-` or `assign-` (respectively). 
We can use these values to decide what all to show in the panel. 

## Relevant Technical Choices

- Did a nested ternary 🥴 for the rendering of the `add category` toggle. It _feels_ easy to read still and seemed like the most straight forward option. Memoizing it would've taken in a lot of props, could make its own component, it just doesn't feel helpful - mostly just passing a buncha props and abstracting something to make it harder to parse through - plus we've got other PRs open or issues getting worked on that would conflict with that update.
- Opted for an e2e test to check on contributor vs admin capabilities because adjusting the story response in the fixture seems like over kill given we're not doing that for any other karma tests, feels like the wrong time to start. 
- Cleaned out the `sort` for ordering the UI, it doesn't seem necessary. 
- Filtering out taxonomies that users don't have the ability to see (`visibility.show_ui`) and can't assign to, if that filtering leaves the user with 0 taxonomies to see then the panel doesn't render. 
- Added some ids to buttons on hierarchical so that e2e selections are easier
- Added axe test to karma 

## To-do

- [x] - Testing this, seems like my default admin user doesn't have permissions for the other taxonomy options for colors, verticals, etc so those aren't rendering anymore. Wondering if some slugs are off or the wrong casing is occurring somewhere? Had a kinda weird rebase. Want to confirm this isn't a regression on my part. 
- [x] - Follow up with #9034 to let users that can't create terms but can still manage add tags that exist. 
- [ ] - Sadly, the e2e tests have been a time suck so I isolated them in their own ticket: #9149 to unblock this.

## User-facing changes

(Blocked by a feature flag, assuming that flag is on if you're logged in as an admin or an editor you can add new categories, if you're logged in as an author or a contributor you won't see that button anymore)

| Admin, Editor | Author, Contributor |
| --- | --- |
|![Screen Shot 2021-09-21 at 1 51 52 PM](https://user-images.githubusercontent.com/10720454/134246222-1a24ecf9-4469-44a4-80cd-8ce4e412e503.png)|  <img width="366" alt="Screen Shot 2021-09-21 at 1 52 15 PM" src="https://user-images.githubusercontent.com/10720454/134246269-36afe72c-08dd-4afe-9052-30e30e072a7d.png">
|![Screen Shot 2021-09-21 at 1 51 45 PM](https://user-images.githubusercontent.com/10720454/134246224-83396a5c-eb43-492e-b599-730c39a587ed.png) |  <img width="312" alt="Screen Shot 2021-09-21 at 1 52 10 PM" src="https://user-images.githubusercontent.com/10720454/134246337-b6a107cd-2847-4efb-8ad7-05aab4c06072.png"> |

## Testing Instructions

- Enable Taxonomies experiment
- Log into the editor as different roles (our local environments come set up for `admin`, `editor`, `author`, `contributor`, `subscriber` (passwords all default to `password`) and go to taxonomies. Admin and editor roles should be able to add in new categories while authors and contributors should not see that option at all.




This PR can be tested by following these steps:

1.


## Reviews

### Does this PR have a security-related impact?



### Does this PR change what data or activity we track or use?



### Does this PR have a legal-related impact?



## Checklist




---



Fixes #9011
